test/extensions/stats_sinks/hystrix/hystrix_test.cc

```diff
@@ -0,0 +1,360 @@
+  // Register callback to sink.
+  sink_->registerConnection(&callbacks_);
+  sink_->flush(source_);
+  std::string data_message = TestUtility::bufferToString(buffer);
+  validate_results(data_message, 0, 0, 0, 0, 0, window_size_);
+}
+
+TEST_F(HystrixSinkTest, BasicFlow) {
+  InSequence s;
+  Buffer::OwnedImpl buffer = createClusterAndCallbacks();
+  // Register callback to sink.
+  sink_->registerConnection(&callbacks_);
+
+  // Arbitrary numbers for testing. Make sure error > timeout.
+  uint64_t success_step = 44;
+  uint64_t error_step = 33;
+  uint64_t timeout_step = 22;
+  uint64_t rejected_step = 11;
+
+  for (uint64_t i = 0; i < (window_size_ + 1); i++) {
+    buffer.drain(buffer.length());
+    ON_CALL(cluster1_.error_5xx_counter_, value()).WillByDefault(Return((i + 1) * error_step));
+    ON_CALL(cluster1_.success_counter_, value()).WillByDefault(Return((i + 1) * success_step));
+    cluster1_.cluster_info_->stats().upstream_rq_timeout_.add(timeout_step);
+    cluster1_.cluster_info_->stats().upstream_rq_pending_overflow_.add(rejected_step);
+    sink_->flush(source_);
+  }
+
+  std::string rolling_map = sink_->printRollingWindows();
+  EXPECT_NE(std::string::npos, rolling_map.find(cluster1_name_ + ".total"));
+
+  // Check stream format and data.
+  validate_results(TestUtility::bufferToString(buffer), success_step, error_step, timeout_step, 0,
+                   rejected_step, window_size_);
+
+  // Check the values are reset.
+  buffer.drain(buffer.length());
+  sink_->resetRollingWindow();
+  sink_->flush(source_);
+  validate_results(TestUtility::bufferToString(buffer), 0, 0, 0, 0, 0, window_size_);
+}
+
+//
+TEST_F(HystrixSinkTest, Disconnect) {
+  InSequence s;
+  Buffer::OwnedImpl buffer = createClusterAndCallbacks();
+
+  sink_->flush(source_);
+  EXPECT_EQ(buffer.length(), 0);
+
+  // Register callback to sink.
+  sink_->registerConnection(&callbacks_);
+
+  // Arbitrary numbers for testing. Make sure error > timeout.
+  uint64_t success_step = 1;
+
+  for (uint64_t i = 0; i < (window_size_ + 1); i++) {
+    buffer.drain(buffer.length());
+    ON_CALL(cluster1_.success_counter_, value()).WillByDefault(Return((i + 1) * success_step));
+    sink_->flush(source_);
+  }
+
+  EXPECT_EQ(getStreamField(TestUtility::bufferToString(buffer), "rollingCountSuccess"),
+            std::to_string(success_step * window_size_));
+  EXPECT_NE(buffer.length(), 0);
+
+  // Disconnect.
+  buffer.drain(buffer.length());
+  sink_->unregisterConnection(&callbacks_);
+  sink_->flush(source_);
+  EXPECT_EQ(buffer.length(), 0);
+
+  // Reconnect.
+  buffer.drain(buffer.length());
+  sink_->registerConnection(&callbacks_);
+  ON_CALL(cluster1_.success_counter_, value()).WillByDefault(Return(success_step));
+  sink_->flush(source_);
+  EXPECT_EQ(getStreamField(TestUtility::bufferToString(buffer), "rollingCountSuccess"), "0");
+  EXPECT_NE(buffer.length(), 0);
+}
+
+TEST_F(HystrixSinkTest, AddAndRemoveClusters) {
+  InSequence s;
```

Comment:
This test has improved a lot. Thank you! It is still a bit long.  I'm not sure if it could be broken up into multiple tests. If so, that'd be my first suggestion. I think there's a notion here that the latter part of the test depends on the former, but there's really some distinct tests. here.

It's not obvious whether there are repeating patterns to factor out, but maybe it should be broken into single-call helpers anyway, just for legibility.

A search on this topic yielded https://refactoring.guru/extract-method

One thing that comes to mind is to break this into 2 tests with one shared helper.
```
TEST_F(HysterixSinkTest, AddClusters) {
  AddHelper();
  EXPECT that add worked
}

TEST_F(HystrixSinkTest, RemoveClusters) {
  AddHelper();
  RemoveHelper();
  EXPECT that remove worked.
}
```
I think it's OK to outline RemoveHelper even though it's single use, for brevity and parallelism.

WDYT?

Reply:
Thanks for the link, I learned. 
Separating into two tests - I can do it. But it means that some actions will be done twice in both tests - I want the cache to be dirty, so AddHelper is not just setting up stuff.
In the new version (pushing soon) I attempted to factor out some stuff in the spirit of the link you sent, and I was also able to move away some unnecessary actions from the clusters test. please see how it looks now, and if it still looks two much, I can split into to test as you suggested.

Reply:
I think splitting them up would help readability.

I'm not sure why there's an issue with the cache being dirty though.  Can you explain?

Reply:
The idea is to see that if a cluster that was active is removed, and then a new cluster with same name is added, there are no leftover data from the previous cluster (it was an actual bug I fixed)

Reply:
Sounds good. So can you split it up help readability?  My suggestion does not change flow of the test you have here, just makes it easier to read (and additionally adds a simpler test).